src/txmempool.cpp

```diff
@@ -790,8 +790,39 @@ bool CCoinsViewMemPool::HaveCoins(const uint256 &txid) const {
 
 size_t CTxMemPool::DynamicMemoryUsage() const {
     LOCK(cs);
-    // Estimate the overhead of mapTx to be 6 pointers + an allocation, as no exact formula for boost::multi_index_contained is implemented.
-    return memusage::MallocUsage(sizeof(CTxMemPoolEntry) + 6 * sizeof(void*)) * mapTx.size() + memusage::DynamicUsage(mapNextTx) + memusage::DynamicUsage(mapDeltas) + cachedInnerUsage;
+
+    size_t total = 0;
+
+    // Estimate the overhead of mapTx to be 6 pointers + an allocation, as no exact formula for
+    // boost::multi_index_contained is implemented.
+    total += memusage::MallocUsage(sizeof(CTxMemPoolEntry) + 6 * sizeof(void*)) * mapTx.size();
+
+    // Two metadata maps inherited from Bitcoin Core
+    total += memusage::DynamicUsage(mapNextTx) + memusage::DynamicUsage(mapDeltas);
+
+    // Saves iterating over the full map
+    total += cachedInnerUsage;
+
+    // Wallet notification
+    total += memusage::DynamicUsage(mapRecentlyAddedTx);
+
+    // Nullifier set tracking
+    total += memusage::DynamicUsage(mapSproutNullifiers) + memusage::DynamicUsage(mapSaplingNullifiers);
+
+    // DoS mitigation
+    total += memusage::DynamicUsage(recentlyEvicted) + memusage::DynamicUsage(weightedTxTree);
+
+    // Insight-related structures
+    size_t insight;
+    insight += memusage::MallocUsage(sizeof(CMempoolAddressDeltaKey) +
+        sizeof(CMempoolAddressDelta) + sizeof(CMempoolAddressDeltaKeyCompare)) * mapAddress.size();
```

Comment:
This doesn't compute the correct value; for example, the comparator is not stored in the heap-allocated map elements. Instead, add an extra template parameter to `memusage::DynamicUage(std::map)` so that it works for maps that set a custom comparator:
```diff
diff --git a/src/memusage.h b/src/memusage.h
index e3e60d0a0..98d03fd63 100644
--- a/src/memusage.h
+++ b/src/memusage.h
@@ -88,8 +88,8 @@ static inline size_t DynamicUsage(const std::set<X>& s)
     return MallocUsage(sizeof(stl_tree_node<X>)) * s.size();
 }
 
-template<typename X, typename Y>
-static inline size_t DynamicUsage(const std::map<X, Y>& m)
+template<typename X, typename Y, typename C>
+static inline size_t DynamicUsage(const std::map<X, Y, C>& m)
 {
     return MallocUsage(sizeof(stl_tree_node<std::pair<const X, Y> >)) * m.size();
 }
```

Reply:
Please check https://github.com/zcash/zcash/pull/4306/commits/6197f536860edd582611fc63c8c3a98d587dc5de and let me know, i am not totally sure i understood it correctly. thanks.